File: plug-in/lattigo/acc/check_sig.go
package acc

import (
	"encoding/json"
	"fmt"
	"reflect"
	"strconv"

	"github.com/tuneinsight/lattigo/v4/bfv"
	"github.com/tuneinsight/lattigo/v4/ckks"
	"github.com/tuneinsight/lattigo/v4/rlwe"
)

func check_key_signature(rlk *rlwe.RelinearizationKey, glk *rlwe.RotationKeySet, key_signature map[string]interface{}) {
	if rlk != nil {
		level := int(key_signature["rlk"].(float64))
		if level > rlk.Keys[0].LevelQ() {
			panic(fmt.Sprintf("The relinearization key level %d does not match the signature level %d", rlk.Keys[0].LevelQ(), level))
		}
	}

	if glk != nil {
		for galois_elem_str, level := range key_signature["glk"].(map[string]interface{}) {
			galois_elem, err := strconv.ParseUint(galois_elem_str, 10, 64)
			if err != nil {
				panic(fmt.Sprintf("galois_elem %s is not an integer", galois_elem_str))
			}
			if glk.Keys[uint64(galois_elem)] == nil {
				panic(fmt.Sprintf("The rotation key glk_%d is not prepared", galois_elem))
			}
			if int(level.(float64)) > glk.Keys[uint64(galois_elem)].LevelQ() {
				panic(fmt.Sprintf("The rotation key glk_%d level %d does not match the signature level %d", galois_elem, glk.Keys[uint64(galois_elem)].LevelQ(), int(level.(float64))))
			}
		}
	}
}

func check_with_sig(arg GoVectorArgument, expected_id string, expected_type string, expected_shape []interface{}, expected_level int) {
	if arg.ArgId != expected_id {
		panic(fmt.Sprintf("For argument %s, expected id is %s, but input id is %s.", arg.ArgId, expected_id, arg.ArgId))
	}

	expected_size := 1
	for _, dim := range expected_shape {
		expected_size *= int(dim.(float64))
	}
	if len(arg.Data) != expected_size {
		panic(fmt.Sprintf("For argument %s, expected size is %d, but input size is %d.", arg.ArgId, expected_size, len(arg.Data)))
	}

	if reflect.TypeOf(arg.Data).Kind() != reflect.Slice {
		panic(fmt.Sprintf("For argument %s, type is not a slice.", arg.ArgId))
	}

	for i := range arg.Data {
		operand := arg.Data[i]
		switch operand := operand.(type) {
		case *bfv.Ciphertext:
			if expected_type != "ct" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is ct.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		case *bfv.Plaintext:
			if expected_type != "pt" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is pt.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		case *bfv.PlaintextRingT:
			if expected_type != "pt_ringt" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is pt_ring.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		case *bfv.PlaintextMul:
			if expected_type != "pt_mul" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is pt_mul.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		case *ckks.Ciphertext:
			if expected_type != "ct" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is ct.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		case *ckks.Plaintext:
			if expected_type != "pt" {
				panic(fmt.Sprintf("For argument %s, expected type is %s, but input type is pt.", arg.ArgId, expected_type))
			}
			if expected_level != operand.Level() {
				panic(fmt.Sprintf("For argument %s, expected level is %d, but input level is %d.", arg.ArgId, expected_level, operand.Level()))
			}
		default:
			panic("Unsupported operand type")
		}
	}

}

func check_parameter(params interface{}, param_json map[string]interface{}) {
	if _, ok := param_json["n"]; !ok {
		panic("Parameter JSON missing 'n' field")
	}
	if _, ok := param_json["q"]; !ok {
		panic("Parameter JSON missing 'q' field")
	}

	// Parse N using json.Number
	n_val, err := param_json["n"].(json.Number).Int64()
	if err != nil {
		panic(fmt.Sprintf("Failed to parse n: %v", err))
	}
	expected_n := int(n_val)

	expected_q_interface := param_json["q"].([]interface{})
	expected_q := make([]uint64, len(expected_q_interface))
	for i, v := range expected_q_interface {
		val, err := v.(json.Number).Int64()
		if err != nil {
			panic(fmt.Sprintf("Failed to parse Q[%d]: %v", i, err))
		}
		expected_q[i] = uint64(val)
	}

	switch params := params.(type) {
	case bfv.Parameters:
		actual_n := params.N()
		if actual_n != expected_n {
			panic(fmt.Sprintf("BFV parameter N mismatch: expected %d, got %d", expected_n, actual_n))
		}

		// Check t for BFV
		if t_val, ok := param_json["t"]; ok {
			expected_t, err := t_val.(json.Number).Int64()
			if err != nil {
				panic(fmt.Sprintf("Failed to parse t: %v", err))
			}
			actual_t := params.T()
			if actual_t != uint64(expected_t) {
				panic(fmt.Sprintf("BFV parameter t mismatch: expected %d, got %d", expected_t, actual_t))
			}
		}

		// Check Q moduli
		actual_q := params.Q()

		if len(actual_q) != len(expected_q) {
			panic(fmt.Sprintf("BFV parameter Q count mismatch: expected %d, got %d", len(expected_q), len(actual_q)))
		}

		for i := 0; i < len(expected_q); i++ {
			if actual_q[i] != expected_q[i] {
				panic(fmt.Sprintf("BFV parameter Q[%d] mismatch: expected %d, got %d", i, expected_q[i], actual_q[i]))
			}
		}

		// Check P moduli if specified
		if p_val, ok := param_json["p"]; ok {
			expected_p_interface := p_val.([]interface{})
			expected_p := make([]uint64, len(expected_p_interface))
			for i, v := range expected_p_interface {
				val, err := v.(json.Number).Int64()
				if err != nil {
					panic(fmt.Sprintf("Failed to parse P[%d]: %v", i, err))
				}
				expected_p[i] = uint64(val)
			}

			actual_p := params.P()

			if len(actual_p) != len(expected_p) {
				panic(fmt.Sprintf("BFV parameter P count mismatch: expected %d, got %d", len(expected_p), len(actual_p)))
			}

			for i := 0; i < len(expected_p); i++ {
				if actual_p[i] != expected_p[i] {
					panic(fmt.Sprintf("BFV parameter P[%d] mismatch: expected %d, got %d", i, expected_p[i], actual_p[i]))
				}
			}
		}

	case ckks.Parameters:
		actual_n := params.N()
		if actual_n != expected_n {
			panic(fmt.Sprintf("CKKS parameter N mismatch: expected %d, got %d", expected_n, actual_n))
		}

		// Check Q moduli
		actual_q := params.Q()

		if len(actual_q) != len(expected_q) {
			panic(fmt.Sprintf("CKKS parameter Q count mismatch: expected %d, got %d", len(expected_q), len(actual_q)))
		}

		for i := 0; i < len(expected_q); i++ {
			if actual_q[i] != expected_q[i] {
				panic(fmt.Sprintf("CKKS parameter Q[%d] mismatch: expected %d, got %d", i, expected_q[i], actual_q[i]))
			}
		}

		// Check P moduli if specified
		if p_val, ok := param_json["p"]; ok {
			expected_p_interface := p_val.([]interface{})
			expected_p := make([]uint64, len(expected_p_interface))
			for i, v := range expected_p_interface {
				val, err := v.(json.Number).Int64()
				if err != nil {
					panic(fmt.Sprintf("Failed to parse P[%d]: %v", i, err))
				}
				expected_p[i] = uint64(val)
			}

			actual_p := params.P()

			if len(actual_p) != len(expected_p) {
				panic(fmt.Sprintf("CKKS parameter P count mismatch: expected %d, got %d", len(expected_p), len(actual_p)))
			}

			for i := 0; i < len(expected_p); i++ {
				if actual_p[i] != expected_p[i] {
					panic(fmt.Sprintf("CKKS parameter P[%d] mismatch: expected %d, got %d", i, expected_p[i], actual_p[i]))
				}
			}
		}

	default:
		panic("Unknown parameter type for parameter checking")
	}
}

func check_signatures(param interface{}, rlk *rlwe.RelinearizationKey, glk *rlwe.RotationKeySet, args []GoVectorArgument, task_signature map[string]interface{}, online_phase bool) int {
	switch param.(type) {
	case bfv.Parameters:
		if task_signature["algorithm"].(string) != "BFV" {
			panic(fmt.Sprintf("task_signature algo type %s does not match bfv.Parameters", task_signature["type"].(string)))
		}
	case ckks.Parameters:
		if task_signature["algorithm"].(string) != "CKKS" {
			panic(fmt.Sprintf("task_signature algo type %s does not match ckks.Parameters", task_signature["type"].(string)))
		}
	default:
		panic("param type is not supported")
	}

	check_key_signature(rlk, glk, task_signature["key"].(map[string]interface{}))

	var data_signature []interface{}
	if online_phase {
		data_signature = task_signature["online"].([]interface{})
	} else {
		data_signature = task_signature["offline"].([]interface{})
	}

	n_in_args := 0
	for i, arg := range args {
		expected_id := data_signature[i].(map[string]interface{})["id"].(string)
		expected_type := data_signature[i].(map[string]interface{})["type"].(string)
		expected_shape := data_signature[i].(map[string]interface{})["size"].([]interface{})
		expected_level := int(data_signature[i].(map[string]interface{})["level"].(float64))
		check_with_sig(arg, expected_id, expected_type, expected_shape, expected_level)

		phase := data_signature[i].(map[string]interface{})["phase"].(string)
		if phase == "in" || phase == "offline" {
			n_in_args++
		}
	}

	return n_in_args
}

File: plug-in/SEAL/acc/check_sig.cpp
#include "runner.h"
#include <stdexcept>

static std::map<SealArgumentType, std::string> seal_argument_type_str_map = {
    {SealArgumentType::RELIN_KEY, "rlk"},
    {SealArgumentType::GALOIS_KEY, "glk"},
    {SealArgumentType::PLAINTEXT, "pt"},
    {SealArgumentType::CIPHERTEXT, "ct"},
};

static std::map<std::string, SealArgumentType> str_seal_argument_type_map = {
    {"rlk", SealArgumentType::RELIN_KEY},      {"glk", SealArgumentType::GALOIS_KEY},
    {"pt_ringt", SealArgumentType::PLAINTEXT}, {"pt", SealArgumentType::PLAINTEXT},
    {"ct", SealArgumentType::CIPHERTEXT},
};

static void check_with_sig(const SealVectorArgument& seal_arg,
                           const std::string& expected_id,
                           SealArgumentType expected_type,
                           const std::vector<size_t>& expected_shape,
                           int expected_level) {
    if (seal_arg.arg_id != expected_id) {
        std::string message = "For argument " + seal_arg.arg_id + ", expected id is " + expected_id +
                              ", but input id is " + seal_arg.arg_id + ".";
        throw std::runtime_error(message);
    }

    if (seal_arg.type != expected_type) {
        std::string message = "For argument " + seal_arg.arg_id + ", expected type is " +
                              seal_argument_type_str_map[expected_type] + ", but input type is " +
                              seal_argument_type_str_map[seal_arg.type] + ".";
        throw std::runtime_error(message);
    }

    int expected_size = 1;
    for (auto x : expected_shape) {
        expected_size *= x;
    }
    if (seal_arg.flat_data.size() != expected_size) {
        std::string message = "For argument " + seal_arg.arg_id + ", expected size is " +
                              std::to_string(expected_size) + ", but input size is " +
                              std::to_string(seal_arg.flat_data.size()) + ".";
        throw std::runtime_error(message);
    }

    if (seal_arg.level != expected_level) {
        std::string message = "For argument " + seal_arg.arg_id + ", expected level is " +
                              std::to_string(expected_level) + ", but input level is " +
                              std::to_string(seal_arg.level) + ".";
        throw std::runtime_error(message);
    }
}

static void
check_key_signatures(const seal::RelinKeys& rlk, const seal::GaloisKeys& glk, const nlohmann::json& key_signature) {
    int rlk_level_sig = key_signature["rlk"].get<int>();
    if (rlk_level_sig >= 0) {
        int rlk_level = rlk.key(2)[0].data().coeff_modulus_size() - 1;
        if (rlk_level_sig > rlk_level) {
            throw std::runtime_error("Level of relin key is smaller than the expected level.");
        }
    }

    if (!key_signature["glk"].empty()) {
        for (auto& item : key_signature["glk"].items()) {
            uint64_t gal_el = stoul(item.key());
            int glk_level_sig = item.value().get<int>();
            auto& ksk = glk.key(gal_el);
            int ksk_level = ksk[0].data().coeff_modulus_size() - 1;
            if (glk_level_sig > ksk_level) {
                throw std::runtime_error("Level of Galois key is smaller than the expected level.");
            }
        }
    }
}

void check_parameter(seal::SEALContext* context, const nlohmann::json& param_json) {
    if (!param_json.contains("n")) {
        throw std::runtime_error("Parameter JSON missing 'n' field");
    }
    if (!param_json.contains("q")) {
        throw std::runtime_error("Parameter JSON missing 'q' field");
    }

    int expected_n = param_json["n"].get<int>();
    std::vector<uint64_t> expected_q = param_json["q"].get<std::vector<uint64_t>>();

    auto& params = context->key_context_data()->parms();
    int actual_n = params.poly_modulus_degree();

    if (actual_n != expected_n) {
        throw std::runtime_error("Parameter N mismatch: expected " + std::to_string(expected_n) + ", got " +
                                 std::to_string(actual_n));
    }

    if (params.scheme() == seal::scheme_type::bfv) {
        // Check t for BFV
        if (param_json.contains("t")) {
            uint64_t expected_t = param_json["t"].get<uint64_t>();
            uint64_t actual_t = params.plain_modulus().value();
            if (actual_t != expected_t) {
                throw std::runtime_error("BFV parameter t mismatch: expected " + std::to_string(expected_t) + ", got " +
                                         std::to_string(actual_t));
            }
        }
    }

    // Check Q moduli
    auto& coeff_modulus = params.coeff_modulus();

    // Separate Q and P moduli (last modulus is special prime P)
    int q_count = coeff_modulus.size();
    int p_count = 0;

    if (param_json.contains("p")) {
        std::vector<uint64_t> expected_p = param_json["p"].get<std::vector<uint64_t>>();
        p_count = expected_p.size();
        q_count = coeff_modulus.size() - p_count;

        // Check Q count
        if (q_count != expected_q.size()) {
            throw std::runtime_error("Parameter Q count mismatch: expected " + std::to_string(expected_q.size()) +
                                     ", got " + std::to_string(q_count));
        }

        // Check each Q value
        for (int i = 0; i < expected_q.size(); i++) {
            if (coeff_modulus[i].value() != expected_q[i]) {
                throw std::runtime_error("Parameter Q[" + std::to_string(i) + "] mismatch: expected " +
                                         std::to_string(expected_q[i]) + ", got " +
                                         std::to_string(coeff_modulus[i].value()));
            }
        }

        // Check P count
        if (p_count != expected_p.size()) {
            throw std::runtime_error("Parameter P count mismatch: expected " + std::to_string(expected_p.size()) +
                                     ", got " + std::to_string(p_count));
        }

        // Check each P value
        for (int i = 0; i < expected_p.size(); i++) {
            if (coeff_modulus[q_count + i].value() != expected_p[i]) {
                throw std::runtime_error("Parameter P[" + std::to_string(i) + "] mismatch: expected " +
                                         std::to_string(expected_p[i]) + ", got " +
                                         std::to_string(coeff_modulus[q_count + i].value()));
            }
        }
    } else {
        // No P specified, all moduli are Q
        if (coeff_modulus.size() != expected_q.size()) {
            throw std::runtime_error("Parameter Q count mismatch: expected " + std::to_string(expected_q.size()) +
                                     ", got " + std::to_string(coeff_modulus.size()));
        }

        for (int i = 0; i < expected_q.size(); i++) {
            if (coeff_modulus[i].value() != expected_q[i]) {
                throw std::runtime_error("Parameter Q[" + std::to_string(i) + "] mismatch: expected " +
                                         std::to_string(expected_q[i]) + ", got " +
                                         std::to_string(coeff_modulus[i].value()));
            }
        }
    }
}

int check_signatures(seal::SEALContext* context,
                     const seal::RelinKeys& rlk,
                     const seal::GaloisKeys& glk,
                     const std::vector<SealVectorArgument>& seal_args,
                     const nlohmann::json& task_sig_json,
                     bool online_phase) {
    int algo;
    if (context->key_context_data()->parms().scheme() == seal::scheme_type::bfv) {
        algo = 0;
        if (task_sig_json["algorithm"] != "BFV") {
            throw std::runtime_error("algo error");
        }
    } else if (context->key_context_data()->parms().scheme() == seal::scheme_type::ckks) {
        algo = 1;
        if (task_sig_json["algorithm"] != "CKKS") {
            throw std::runtime_error("algo error");
        }
    } else {
        throw std::runtime_error("context type error");
    }

    check_key_signatures(rlk, glk, task_sig_json["key"]);

    auto data_sig_json = online_phase ? task_sig_json["online"].get<std::vector<nlohmann::json>>() :
                                        task_sig_json["offline"].get<std::vector<nlohmann::json>>();

    int n_in_args = 0;

    for (int i = 0; i < seal_args.size(); i++) {
        std::string expected_id = data_sig_json[i]["id"].get<std::string>();
        SealArgumentType expected_type = str_seal_argument_type_map[data_sig_json[i]["type"].get<std::string>()];
        std::vector<uint64_t> expected_shape = data_sig_json[i]["size"].get<std::vector<uint64_t>>();
        int expected_level = data_sig_json[i]["level"].get<int>();
        check_with_sig(seal_args[i], expected_id, expected_type, expected_shape, expected_level);

        std::string phase = data_sig_json[i]["phase"].get<std::string>();
        if (phase == "in" or phase == "offline") {
            n_in_args++;
        }
    }

    return n_in_args;
}
